app/custom_auth.py

```diff
@@ -0,0 +1,74 @@
+from django.contrib import auth
+from django.conf import settings
+from django.contrib.auth.models import User
+
+class CustomSSOMiddleware:
+    """
+    A custom middleware that allows SSO via HTTP header passed in the request
+    The name of HTTP header is defined in the settings.py by the variable HEADER_CUSTOM_USER_NAME  
+
+    
+    (1) If HTTP header with the user name is not passed no further processing is done by the middleware.
+    (2) If HTTP header with the user name is passed then:
+        (2.1) Authenciate the user if already exist in the database. 
+        (2.2) Create user if does not exist in the database and then autenicate the user. 
+        (2.3) User is now signed in and session cookies are set.  
+    ...
+    """
+    def __init__(self, get_response):
+        print ('LOG: Inside the custom middleware!   __init__')
+        self.get_response = get_response
+     
+    def __call__(self, request):
+        print ('LOG: Inside the custom middleware!   __call__')
+        
+        header =  settings.HEADER_CUSTOM_USER_NAME
+        username= ""
+
+        try:
+            username = request.META[header]
+        except:
+         
+            print ("LOG: User name header not found: " + header)
+            response = self.get_response(request)
+            return response
+
+        print( 'LOG: Username provided by the header :' + username)
+        is_user_exists = username_present(username)
+        user = None
+
+        if is_user_exists == True:
+            user = auth.authenticate(remote_user=username)
+        else:
+            print ('LOG: User is not defined: '+username)
+            if settings.CREATE_USER == 'TRUE':
+                user = create_user(username)
+                if user is not None:
+                   user = auth.authenticate(remote_user=username)
+
+        if user:
+            request.user = user
+            auth.login(request, user)
+
+        response = self.get_response(request)
+
+        return response
+
+def username_present(username):
+    if User.objects.filter(username=username).exists():
+        return True
+
+def create_user(username):
+
+    print ('LOG: create_user called: ' +username)
+    user = None
+    try:
+        email = username + "@noreply.com"
+        user = User.objects.create_user(username=username,email=email)
+        user.is_superuser=settings.IS_SUPER
+        user.save()
+
+        return user
+    except:
+        print ('User creation failed: '+ username)
+        return None
```

Comment:
We should probably return something more specific like a `HttpResponseForbidden`.

Reply:
I have added ```PermissionDenied``` to send the 403 back as the response.  
This explains [PermissionDenied](https://stackoverflow.com/questions/6618939/how-do-i-raise-a-response-forbidden-in-django) usage and it seems to be the better option over ```HttpResponseForbidden```

Following is the response when user creation failed. 

```
$  curl -I "http://localhost:8000" -H "X-Custom-User-Name: guest0019"
HTTP/1.1 403 Forbidden
Date: Tue, 09 Jul 2019 17:37:13 GMT
Server: WSGIServer/0.2 CPython/3.7.3
Content-Type: text/html
X-Frame-Options: SAMEORIGIN
Content-Length: 22
Vary: Cookie
```